datafusion/src/physical_optimizer/repartition.rs

```diff
@@ -36,33 +128,70 @@ impl Repartition {
     }
 }
 
+/// Recursively visits all `plan`s puts and then optionally adds a
```

Comment:
For my own understanding I'm going to write out what this does.

It does a depth first scan of the tree, and repartitions any plan that:

* Has less than the desired number of partitions
* Has a direct parent that `benefits_from_input_partitioning`
* Does not have a parent that `relies_on_input_order` unless there is an intervening node that does not `maintain_input_order`

Reply:
> Has a direct parent that benefits_from_input_partitioning

I think this is `Has any parent that `benefits_from_input_partitioning` unless there is an intervening node

otherwise yes. I will add this summary as a comment. Thank you